BOLT#02: clarify coop close requirements (#970)
lightning/bolts: 02-peer-protocol.md

  - if it hasn't sent a `funding_created` (if it is a funder) or a `funding_signed` (if it is a fundee):
    - MUST NOT send a `shutdown`
  - MAY send a `shutdown` before a `funding_locked`, i.e. before the funding transaction has reached `minimum_depth`.

>   - if there are uncommitted updates pending on the receiving node's commitment transaction:

Reviewer: @vincenzopalazzo started looking at this in our code and I kinda realized....why do we have this requirement at all? Like, who really cares what's pending when a shutdown goes out? *after* shutdown goes out you shouldn't add anything new, and certainly sending a shutdown out while an update_add is pending is kinda dumb (the peer should reject at that point), but its not like it should break anything, only pre-closing_signed is a real concern here.

Reviewer: I assume a part of the reason for this is so that the receiving end can enforce the "no new update_adds" requirement when receiving `commitment_signed` and not later? Does anyone do this (maybe c-lightning @rustyrussell ?)

Reviewer 2: OK, this was good chatting, and I'm sorry if this came out of my head after the meeting, but I just drafted the lnprototest PR https://github.com/rustyrussell/lnprototest/pull/37 after the meeting and I was able to make some consideration after writing the code.

So, my idea that I shared with @TheBlueMatt is why we care about what we don't receive, and we can be more precise in the fact of what we must not receive.

This will make the spec also less ambiguous, for instance, some things like that
```suggestion
  - if `add_htlc` message is received from the received after shutdown, the sender must fail the htlc 
```

P.S: this make also the lnprototest integration test easy with a single assert `MustNodeReceive('add_htlc')` :)

Reviewer 2: The only reason to have a "black list" is that during the time this black list can blow up?

    - MUST NOT send a `shutdown`.
  - MUST NOT send an `update_add_htlc` after a `shutdown`.

>   - if neither side has a dangling commitment and neither commitment contains any HTLCs (including dust HTLCs):

Reviewer: I don't think "dangling" is super well-defined here, and I'm not convinced we want to make that change anyway.

Reviewer: Hmm, I'd argue there is still "an HTLC in (one of the) counterparty commitment transactions" at that point - it just so happens there are two counterparty commitment transactions until you get the RAA. Still, in either case, this should use more concrete language IMO.

the sender always sends a `commitment_signed` first.

As shutdown implies a desire to terminate, it implies that no new

> HTLCs will be added or accepted.  Once any HTLCs are cleared, there are no dangling commitments and all updates are included on both commitment transactions, the peer

Reviewer: Instead of using the word "dangling" can we specify what we mean clearly - something about commitments for which no `revoke_and_ack` has been sent?

Also, nit: please wrap lines at 80 chars, or so.

may immediately begin closing negotiation, so we ban further updates
to the commitment transaction (in particular, `update_fee` would be

> possible otherwise). However, while there are HTLCs on the commitment transaction, the initiator may find it desirable to increase the feerate as there may be pending HTLCs on the commitment which could timeout.

Reviewer 2: ```suggestion
possible otherwise). However, while there are HTLCs on the commitment 
transaction, the initiator may find it desirable to increase 
the feerate as there may be pending HTLCs on the commitment which could timeout.
```

### Closing Negotiation: `closing_signed`

> Once shutdown is complete, the channel is empty of HTLCs, there are no dangling commitments and all updates are included on both commitments, the final

Reviewer 2: ```suggestion
Once shutdown is complete, the channel is empty of HTLCs, there are no dangling 
commitments and all updates are included on both commitments, the final
```

The same comment of Matt here, be more specific it is good here because we avoid inconsistency for future updates or ambiguity.

    - MUST NOT send a `shutdown`.
  - MUST NOT send multiple `shutdown` messages.
  - MUST NOT send an `update_add_htlc` after a `shutdown`.

>   - if neither side has a commitment for which a revocation is owed and neither
>   commitment contains any HTLCs (including dust HTLCs):

Reviewer 2: ```suggestion
  - if neither side has a commitment for which a revocation is owed and neither
    commitment contains any HTLCs (including dust HTLCs):
```

>   - if neither side has a commitment for which a revocation is owed and neither
>   commitment contains any HTLCs (including dust HTLCs):
>     - MUST NOT send any `update` message after a `shutdown`.

Reviewer 3: Let me know what you think about this proposal, I personally find the intent clearer this way (I find the requirement below your change unclear, since it's somewhat unrelated to when we sent `shutdown`):

```suggestion
  - if no HTLCs remain in either commitment transaction (including dust HTLCs)
    and neither side has a pending `revoke_and_ack` to send:
    - MUST NOT send any `update` message after that point.
```

Author: I agree

Author: done

the sender always sends a `commitment_signed` first.

As shutdown implies a desire to terminate, it implies that no new

> HTLCs will be added or accepted.  Once any HTLCs are cleared, there are no commitments

Reviewer 2: ```suggestion
HTLCs will be added or accepted. Once any HTLCs are cleared, there are no commitments
```
